src/raftstore/store/snap.rs

```diff
@@ -618,6 +670,1025 @@ mod v1 {
+                    continue;
+                }
+                let left = (cf_file.size - cf_file.written_size) as usize;
+                if left == 0 {
+                    self.cf_index += 1;
+                    continue;
+                }
+                let mut file = cf_file.file.as_mut().unwrap();
+                let mut digest = cf_file.write_digest.as_mut().unwrap();
+                if next_buf.len() > left {
+                    try!(file.write_all(&next_buf[0..left]));
+                    digest.write(&next_buf[0..left]);
+                    cf_file.written_size += left as u64;
+                    self.cf_index += 1;
+                    next_buf = &next_buf[left..];
+                } else {
+                    try!(file.write_all(next_buf));
+                    digest.write(next_buf);
+                    cf_file.written_size += next_buf.len() as u64;
+                    return Ok(buf.len());
+                }
+            }
+            let n = buf.len() - next_buf.len();
+            Ok(n)
+        }
+
+        fn flush(&mut self) -> io::Result<()> {
+            if let Some(cf_file) = self.cf_files.get_mut(self.cf_index) {
+                let file = cf_file.file.as_mut().unwrap();
+                try!(file.flush());
+            }
+            Ok(())
+        }
+    }
+
+    impl Drop for Snap {
+        fn drop(&mut self) {
+            if self.cf_files.iter().any(|cf_file| file_exists(&cf_file.tmp_path)) ||
+               file_exists(&self.meta_file.tmp_path) {
+                self.try_delete(true)
+            }
+        }
+    }
+
+    #[cfg(test)]
+    pub mod test {
+        use std::io;
+        use std::sync::{Arc, RwLock};
+        use std::sync::atomic::AtomicUsize;
+        use tempdir::TempDir;
+        use kvproto::metapb::{Peer, Region};
+        use kvproto::raft_serverpb::RaftSnapshotData;
+        use rocksdb::DB;
+
+        use storage::ALL_CFS;
+        use util::{rocksdb, HandyRwLock};
+        use raftstore::Result;
+        use raftstore::store::keys;
+        use raftstore::store::engine::{Snapshot as DbSnapshot, Mutable, Peekable, Iterable};
+        use raftstore::store::peer_storage::JOB_STATUS_RUNNING;
+        use super::super::{SNAP_GEN_PREFIX, need_to_pack, SnapKey, Snapshot};
+        use super::{Snap, BuildContext, ApplyOptions};
+
+        const TEST_STORE_ID: u64 = 1;
+        const TEST_KEY: &[u8] = b"akey";
+
+        pub fn get_test_empty_db(path: &TempDir) -> Result<Arc<DB>> {
+            let p = path.path().to_str().unwrap();
+            let db = try!(rocksdb::new_engine(p, ALL_CFS));
+            Ok(Arc::new(db))
+        }
+
+        pub fn get_test_db(path: &TempDir) -> Result<Arc<DB>> {
+            let p = path.path().to_str().unwrap();
+            let db = try!(rocksdb::new_engine(p, ALL_CFS));
+            let key = keys::data_key(TEST_KEY);
+            // write some data into each cf
+            for (i, cf) in ALL_CFS.into_iter().enumerate() {
```

Comment:
into_iter will move out the value in ALL_CFS, is it right here? can iter not work?